Src/MakingSense.SafeBrowsing.Tests/CanonicalUrlTests.cs

```diff
@@ -0,0 +1,137 @@
+using Assert = MakingSense.SafeBrowsing.Tests.XUnitAssert;
+using TestCase = Xunit.InlineDataAttribute;
+#else
+using NUnit.Framework;
+#endif
+
+namespace MakingSense.SafeBrowsing.Tests
+{
+    [TestFixture]
+    public class CanonicalUrlTests
+    {
+
+#if DNXCORE50
+        [Theory]
+#endif
+        [TestCase("http://host/%25%32%35", "http://host/%25")]
+        [TestCase("http://host/%25%32%35%25%32%35", "http://host/%25%25")]
+        [TestCase("http://host/%2525252525252525", "http://host/%25")]
+        [TestCase("http://host/asdf%25%32%35asd", "http://host/asdf%25asd")]
+        [TestCase("http://host/%%%25%32%35asd%%", "http://host/%25%25%25asd%25%25")]
+        [TestCase("http://www.google.com/", "http://www.google.com/")]
+        [TestCase("http://%31%36%38%2e%31%38%38%2e%39%39%2e%32%36/%2E%73%65%63%75%72%65/%77%77%77%2E%65%62%61%79%2E%63%6F%6D/", "http://168.188.99.26/.secure/www.ebay.com/")]
+        [TestCase("http://195.127.0.11/uploads/%20%20%20%20/.verify/.eBaysecure=updateuserdataxplimnbqmn-xplmvalidateinfoswqpcmlx=hgplmcx/", "http://195.127.0.11/uploads/%20%20%20%20/.verify/.eBaysecure=updateuserdataxplimnbqmn-xplmvalidateinfoswqpcmlx=hgplmcx/")]
+        [TestCase("http://host%23.com/%257Ea%2521b%2540c%2523d%2524e%25f%255E00%252611%252A22%252833%252944_55%252B", "http://host%23.com/~a!b@c%23d$e%25f^00&11*22(33)44_55+")]
+        [TestCase("http://3279880203/blah", "http://195.127.0.11/blah")]
+        [TestCase("http://www.google.com/blah/..", "http://www.google.com/")]
+        [TestCase("www.google.com/", "http://www.google.com/")]
+        [TestCase("www.google.com", "http://www.google.com/")]
+        [TestCase("http://www.evil.com/blah#frag", "http://www.evil.com/blah")]
+        [TestCase("http://www.GOOgle.com/", "http://www.google.com/")]
+        [TestCase("http://www.google.com.../", "http://www.google.com/")]
+        [TestCase("http://www.google.com/foo\tbar\rbaz\n2", "http://www.google.com/foobarbaz2")]
+        [TestCase("http://www.google.com/q?", "http://www.google.com/q?")]
+        [TestCase("http://www.google.com/q?r?", "http://www.google.com/q?r?")]
+        [TestCase("http://www.google.com/q?r?s", "http://www.google.com/q?r?s")]
+        [TestCase("http://evil.com/foo#bar#baz", "http://evil.com/foo")]
+        [TestCase("http://evil.com/foo;", "http://evil.com/foo;")]
+        [TestCase("http://evil.com/foo?bar;", "http://evil.com/foo?bar;")]
+        [TestCase("http://\x01\x80.com/", "http://%01%80.com/")]
+        [TestCase("http://notrailingslash.com", "http://notrailingslash.com/")]
+        [TestCase("http://www.gotaport.com:1234/", "http://www.gotaport.com/")]
+        [TestCase("  http://www.google.com/  ", "http://www.google.com/")]
+        [TestCase("http:// leadingspace.com/", "http://%20leadingspace.com/")]
+        [TestCase("http://%20leadingspace.com/", "http://%20leadingspace.com/")]
+        [TestCase("%20leadingspace.com/", "http://%20leadingspace.com/")]
```

Comment:
Is this fine? should leading space not be trimmed?

Reply:
Yes, these test cases are provided by Google [here](https://developers.google.com/safe-browsing/v4/urls-hashing#canonicalization) to test canonicalization.